packages/dds/sequence-sequential-id/src/sharedStringWithSequentialId.ts

```diff
@@ -0,0 +1,88 @@
+                const set = new SortedSegmentSet();
+                this.walkSegments((segment) => {
+                    if (Marker.is(segment) && segment.getId() !== undefined) {
+                        // If it has an id then we have acked it
+                        set.addOrUpdate(segment);
+                    }
+                    return true;
+                });
+                return set;
+            });
+            this.on("maintenance", this.applyIdToLocalAckedSegment);
+            this.on("sequenceDelta", this.applyIdToRemoteAckedSegment);
+        }
+
+        public static getFactory() {
+            return new SharedStringWithSequentialIdFactory();
+        }
+
+        private readonly applyIdToRemoteAckedSegment = (event: SequenceDeltaEvent): void => {
+            if (event.isLocal) {
+                // Do not apply id for local changes
+                return;
+            }
+
+            event.ranges.forEach((range) => {
+            const markerSegment = range.segment;
+                if (Marker.is(markerSegment)) {
+                    if (event.deltaOperation === MergeTreeDeltaType.INSERT) {
+                        this.applyIdToMarker(markerSegment);
+                    } else if (event.deltaOperation === MergeTreeDeltaType.REMOVE) {
+                        this.sortedMarkers.value.remove(markerSegment);
+                    }
+                }
+            });
+        };
+
+        private readonly applyIdToLocalAckedSegment = (event: SequenceMaintenanceEvent): void => {
+            if (event.deltaArgs.operation === MergeTreeMaintenanceType.ACKNOWLEDGED) {
+                event.ranges.forEach((range) => {
+                    const markerSegment = range.segment;
+                    if (Marker.is(markerSegment)) {
+                        if (event.opArgs?.op.type === MergeTreeDeltaType.INSERT) {
+                            this.applyIdToMarker(markerSegment);
+                        } else if (event.opArgs?.op.type === MergeTreeDeltaType.REMOVE) {
+                            this.sortedMarkers.value.remove(markerSegment);
+                        }
+                    }
+                });
+            }
+        };
+
+        private readonly applyIdToMarker = (markerSegment: Marker): void => {
+            this.sortedMarkers.value.addOrUpdate(markerSegment);
+            const markerItems = this.sortedMarkers.value.items;
+            const newMarkerIndex = markerItems.indexOf(markerSegment);
+            const previousMarkerIndex = newMarkerIndex - 1;
+            const nextMarkerIndex = newMarkerIndex + 1;
+            const previousMarker = previousMarkerIndex >= 0 ?
+             markerItems[previousMarkerIndex] as Marker : undefined;
+            const nextMarker = nextMarkerIndex < markerItems.length ?
+             markerItems[nextMarkerIndex] as Marker : undefined;
+            const previousId: string = previousMarker !== undefined ? previousMarker.getId() ?? digitZero : digitZero;
```

Comment:
I believe previousMarker.getId()  has to have ID. If not, then we have some invariant broken and IDs would be wrong, no? Same on next line

Reply:
It does have to have an id. I think it is just strictNullCheck that because the getId() method returns string | undefined it complains if i don't actually check for it. I can add assert but this should never be undefined due to our code

Reply:
Added asserts to validate previous and next marker have id if defined